### PR TITLE
fix(#8478): Added validPhone validation so that triggers are not fired for patients with invalid phone.

### PR DESCRIPTION
**Pull request by PrjShrestha**
Bug Fix (#8478 ): Invalid phone for patient still running triggers

# Description
Added validPhone validation so that phone validation is done before registration triggers are run. So now if validPhone validation fails for a patient none of the registration triggers configured for the forms are run similar to how Age validation of 0 to 125 failure causes none of the triggers to run if age is above 125 or negative.

Why this was necessary? 
We had a form to add patient with phone number but although patient wasn't added due to invalid phone assign_schedules trigger for the supplied doc was still running so schedules were created for a patient that was invalid. 

Fork Branch: [valid_phone_8478](https://github.com/PrjShrestha/cht-core/tree/valid_phone_8478)

# Code review checklist
- [ x ] Documented: [Here ](https://github.com/medic/cht-docs/pull/1158)
- [ x] Tested: Added validation unit tests
- [x ] Internationalised: All user facing text (Can be configured via translations)
- [x ] Backwards compatible: Yes, does not hamper any preexisting functionality

# Compose URLs

If Build CI hasn't passed, these may 404:

* __CHT_CORE_COMPOSE_URL__
* __COUCH_SINGLE_COMPOSE_URL__
* __COUCH_CLUSTER_COMPOSE_URL__

# License

The software is provided under AGPL-3.0. Contributions to this project are accepted under the same license.

### CODE REVIEW

File: api/resources/translations/messages-en.properties [hunk 1]
```diff
@@ -842,6 +842,7 @@ messages.f.report_accepted_parent = {{patient_name}} ({{patient_id}}) was flagge
 messages.generic.no_provided_patient_id = External ID number support is configured, but is not found on document in the configured location.
 messages.generic.provided_patient_id_not_unique = The provided ID number is already used.
 messages.generic.provided_phone_not_valid = The provided phone is not valid.
+messages.invalid_phone= The provided phone is not valid.
```
Comment (yrimal) [inline api/resources/translations/messages-en.properties:845]: These two have the same content. Can't we use the same key whereve this is called?

File: api/resources/translations/messages-ne.properties [hunk 2]
```diff
@@ -766,6 +766,7 @@ messages.f.report_accepted_parent =
 messages.generic.no_provided_patient_id = 
 messages.generic.provided_patient_id_not_unique = तपाइँले पठाउनुभएको आईडी पहिल्यै प्रयोगमा छ।  
 messages.generic.provided_phone_not_valid = तपाइँले पठाउनुभएको फोन नम्बर मिलेन । फेरी प्रयास गर्नुहोस्।
+messages.invalid_phone= तपाइँले पठाउनुभएको फोन नम्बर मिलेन । फेरी प्रयास गर्नुहोस्।
```
Comment (yrimal) [inline api/resources/translations/messages-ne.properties:769]: Same value for these two, can we use same key ?

File: shared-libs/validation/test/validations.js
Comment (yrimal) [inline shared-libs/validation/test/validations.js:1]: Nice test. Can you add one more passing pure text in phone number?

File: shared-libs/validation/test/validations.js [hunk 3]
```diff
@@ -278,6 +278,100 @@ describe('validations', () => {
     });
   });
 
+
+  it('invalid phone validation should fail if invalid phone is provided', () => {
```
Comment (derickl) [inline shared-libs/validation/test/validations.js:282]:
```suggestion
  it('phone validation should fail if invalid phone is provided', () => {
```

File: shared-libs/validation/test/validations.js [hunk 4]
```diff
@@ -278,6 +278,100 @@ describe('validations', () => {
...
+    const doc = {
+      _id: 'invalid',
+      fields: { phone_number: '+977984111'}
+    };
+    return validation.validate(doc, validations).then(errors => {
+      assert.equal(errors.length, 1);
+    });
+  });
+
+  it('Valid phone validation should pass if valid phone is provided', () => {
```
Comment (derickl) [inline shared-libs/validation/test/validations.js:313]:
```suggestion
  it('phone validation should pass if valid phone number is provided', () => {
```

File: shared-libs/validation/test/validations.js [hunk 5]
```diff
@@ -278,6 +278,100 @@ describe('validations', () => {
...
+    const doc = {
+      _id: 'invalid',
+      fields: { phone_number: '9841134532' }
+    };
+    return validation.validate(doc, validations).then(errors => {
+      assert.equal(errors.length, 0);
+    });
+  });
+
+  it('Valid phone validation should pass if valid phone with extension is provided', () => {
```
Comment (derickl) [inline shared-libs/validation/test/validations.js:344]:
```suggestion
  it('phone validation should pass if valid phone number with country code is provided', () => {
```

File: api/resources/translations/messages-en.properties (hunk 1 above)
Comment (PrjShrestha) [reply api/resources/translations/messages-en.properties:845]: I saw in your app settings you had added invalid_phone hence I thought it was used somewhere. (i.e. Became a sheep) Will remove it.

File: api/resources/translations/messages-ne.properties (hunk 2 above)
Comment (PrjShrestha) [reply api/resources/translations/messages-ne.properties:769]: I saw in your app settings you had added invalid_phone hence I thought it was used somewhere. (i.e. Became a sheep) Will remove it.

File: api/resources/translations/messages-ne.properties [hunk 6]
```diff
@@ -765,7 +765,8 @@ messages.f.report_accepted =
 messages.f.report_accepted_parent = 
 messages.generic.no_provided_patient_id = 
 messages.generic.provided_patient_id_not_unique = तपाइँले पठाउनुभएको आईडी पहिल्यै प्रयोगमा छ।  
-messages.generic.provided_phone_not_valid = तपाइँले पठाउनुभएको फोन नम्बर मिलेन । फेरी प्रयास गर्नुहोस्।
+messages.generic.provided_phone_not_valid = तपाइँले पठाउनुभएको फोन नम्बर मिलेन। कृपया पुनः प्रयास गर्नुहोस्।
+messages.invalid_phone= तपाइँले पठाउनुभएको फोन नम्बर मिलेन। कृपया पुनः प्रयास गर्नुहोस्।
```
Comment (garethbowen) [inline api/resources/translations/messages-ne.properties:769]: I think you removed this key from the en translation file so it should also be removed here?

**Review by garethbowen — CHANGES_REQUESTED** [review]
One minor comment inline.

File: api/resources/translations/messages-ne.properties (hunk 6 above)
Comment (PrjShrestha) [reply api/resources/translations/messages-ne.properties]: @garethbowen Thanks Fixed.

**Review by garethbowen — APPROVED** [review]
Awesome, thanks!

**Pull request merged**